File: swiftPrep/Sources/Sorting/CountingSort.swift

/**
	Sort the sequence in ascending order.

	Time: O(n+k) where k is the range (max - min).
	Space: O(k)
 */
func countingSort(sequence: inout [Int]) {
	guard sequence.count > 1 else { return }

	let minNumber = sequence.min()!
	let maxNumber = sequence.max()! + 1

	var frequencies = Array<Int>(repeating: 0, count: maxNumber - minNumber)

	for number in sequence {
		frequencies[number - minNumber] += 1
	}

	var index = (0..<sequence.endIndex).makeIterator()

	for number in minNumber..<maxNumber {
		let frequency = frequencies[number - minNumber]

		for _ in 0..<frequency {
			guard let index = index.next() else { break }
			sequence[index] = number
		}
	}
}

File: python_prep/sources/sorting/counting_sort.py
def counting_sort(sequence):
	""" 
	Sort the sequence in ascending order. 

	Time: O(n+k) where k is the range 
	Space: O(n+k)

	args:
		sequence: sequence to sort
	"""
	if len(sequence) <= 1:
		return 

	contains_str = type(sequence[0]) is str
	if contains_str:
		for i in range(len(sequence)):
			sequence[i] = ord(sequence[i])

	min_num = min(sequence)
	max_num = max(sequence) + 1

	frequencies = [ 0 for _ in range(min_num, max_num) ]

	for num in sequence:
		frequencies[ num - min_num ] += 1
	
	# StopIteration error should not occur
	index = iter(range(len(sequence)))

	for num in range(min_num, max_num):
		frequency = frequencies[ num - min_num ]

		for _ in range(frequency):
			sequence[next(index)] = num

	if contains_str:
		for i in range(len(sequence)):
			sequence[i] = chr(sequence[i])
